viskores/exec/CellLocatorTwoLevel.h

```diff
@@ -211,72 +210,93 @@ class VISKORES_ALWAYS_EXPORT CellLocatorTwoLevel
+      return viskores::ErrorCode::CellNotFound;
+    }
+    else
     {
-      parametric = pc;
+      cellId = cellIdVec[0];
+      pCoords = pCoordsVec[0];
       return viskores::ErrorCode::Success;
     }
+  }
 
-    return viskores::ErrorCode::CellNotFound;
+  /// @copydoc viskores::exec::CellLocatorUniformGrid::CountAllCells
+  VISKORES_EXEC viskores::IdComponent CountAllCells(const viskores::Vec3f& point) const
+  {
+    viskores::Vec<viskores::Id, 1> cellIdVec = { -1 };
+    viskores::Vec<viskores::Vec3f, 1> pCoordsVec;
+    LastCell lastCell;
+
+    return this->IterateLeaves(point, IterateMode::CountAll, cellIdVec, pCoordsVec, lastCell);
   }
 
-  VISKORES_EXEC
-  viskores::ErrorCode PointInLeaf(const FloatVec3& point,
-                                  const viskores::Id& leafIdx,
-                                  viskores::Id& cellId,
-                                  FloatVec3& parametric) const
+  /// @copydoc viskores::exec::CellLocatorUniformGrid::FindAllCells
+  template <typename CellIdsType, typename ParametricCoordsVecType>
+  VISKORES_EXEC viskores::ErrorCode FindAllCells(const viskores::Vec3f& point,
+                                                 CellIdsType& cellIdVec,
+                                                 ParametricCoordsVecType& pCoordsVec) const
   {
-    viskores::Id start = this->CellStartIndex.Get(leafIdx);
-    viskores::Id end = start + this->CellCount.Get(leafIdx);
+    viskores::IdComponent n = cellIdVec.GetNumberOfComponents();
+    if (pCoordsVec.GetNumberOfComponents() != n)
+      return viskores::ErrorCode::InvalidNumberOfIndices;
 
-    for (viskores::Id i = start; i < end; ++i)
-    {
-      viskores::Vec3f pc;
+    if (n == 0)
+      return viskores::ErrorCode::Success;
 
-      viskores::Id cid = this->CellIds.Get(i);
-      if (this->PointInCell(point, cid, pc) == viskores::ErrorCode::Success)
-      {
-        cellId = cid;
-        parametric = pc;
-        return viskores::ErrorCode::Success;
-      }
-    }
+    for (viskores::IdComponent i = 0; i < n; i++)
+      cellIdVec[i] = -1;
 
-    return viskores::ErrorCode::CellNotFound;
+    LastCell lastCell;
+    viskores::IdComponent nCells =
+      this->IterateLeaves(point, IterateMode::FindAll, cellIdVec, pCoordsVec, lastCell);
+    VISKORES_ASSERT(n == nCells);
+    if (nCells == 0)
+      return viskores::ErrorCode::CellNotFound;
+
+    return viskores::ErrorCode::Success;
   }
 
+private:
+  // Shared leaf traversal: modes are FindOne, CountAll, FindAll
+  enum class IterateMode
+  {
+    FindOne,
+    CountAll,
+    FindAll
+  };
 
-  VISKORES_EXEC
-  viskores::ErrorCode FindCellImpl(const FloatVec3& point,
-                                   viskores::Id& cellId,
-                                   FloatVec3& parametric,
-                                   LastCell& lastCell) const
+  template <typename CellIdVecType, typename ParametricCoordsVecType>
+  VISKORES_EXEC viskores::Id IterateLeaves(const FloatVec3& point,
```

Comment:
Probably better to return a `viskores::IdComponent`.

```suggestion
  VISKORES_EXEC viskores::IdComponent IterateLeaves(const FloatVec3& point,
```